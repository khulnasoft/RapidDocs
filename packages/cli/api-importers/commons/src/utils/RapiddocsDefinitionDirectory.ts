import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { RelativeFilePath, sep } from "@khulnasoft/path-utils";

export class RapiddocsDefinitionDirectory {
    private files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};
    private directories: Record<string, RapiddocsDefinitionDirectory> = {};

    public getAllFiles(): Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> {
        const files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};

        const walk = (root: RapiddocsDefinitionDirectory, currentPath?: string) => {
            for (const [relativeFilePath, definition] of Object.entries(root.files)) {
                const fullRelativeFilePath =
                    currentPath != null
                        ? RelativeFilePath.of(`${currentPath}${sep}${relativeFilePath}`)
                        : RelativeFilePath.of(relativeFilePath);
                files[fullRelativeFilePath] = definition;
            }
            const sortedDirectories = Object.keys(root.directories).sort();
            for (const directory of sortedDirectories) {
                const nextPath = currentPath != null ? `${currentPath}${sep}${directory}` : directory;
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const nextDirectory = root.directories[directory]!;
                walk(nextDirectory, nextPath);
            }
        };

        walk(this);

        return files;
    }

    public getOrCreateFile(relativeFilePath: RelativeFilePath): RawSchemas.DefinitionFileSchema {
        return this.getOrCreateFileRecursive(relativeFilePath.split(sep));
    }

    private getOrCreateFileRecursive(pathParts: string[]): RawSchemas.DefinitionFileSchema {
        if (pathParts.length === 1) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return (this.files[RelativeFilePath.of(pathParts[0]!)] ??= {});
        }
        const [directory, ...remainingPath] = pathParts;
        if (directory == null) {
            throw new Error(`Internal error; cannot add file with path: ${pathParts}`);
        }
        if (!this.directories[directory]) {
            this.directories[directory] = new RapiddocsDefinitionDirectory();
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.directories[directory]!.getOrCreateFileRecursive(remainingPath);
    }
}
