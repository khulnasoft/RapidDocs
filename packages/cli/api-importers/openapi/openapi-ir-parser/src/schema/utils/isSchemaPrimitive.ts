import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaPrimitive(schema: Schema): boolean {
    if (schema.type === "primitive") {
        return true;
    } else if (schema.type === "nullable" || schema.type === "optional") {
        return isSchemaPrimitive(schema.value);
    }
    return false;
}
