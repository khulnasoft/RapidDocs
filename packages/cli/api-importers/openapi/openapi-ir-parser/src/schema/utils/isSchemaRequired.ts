import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaRequired(schema: Schema): boolean {
    return schema.type !== "optional" && schema.type !== "nullable";
}
