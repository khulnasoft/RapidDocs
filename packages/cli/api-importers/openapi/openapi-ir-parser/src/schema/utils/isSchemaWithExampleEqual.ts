import { SchemaWithExample, isSchemaEqual } from "@khulnasoft/openapi-ir";

import { convertSchemaWithExampleToSchema } from "./convertSchemaWithExampleToSchema";

// only diffs the schema, not the example
export function isSchemaWithExampleEqual(a: SchemaWithExample, b: SchemaWithExample): boolean {
    return isSchemaEqual(convertSchemaWithExampleToSchema(a), convertSchemaWithExampleToSchema(b));
}
