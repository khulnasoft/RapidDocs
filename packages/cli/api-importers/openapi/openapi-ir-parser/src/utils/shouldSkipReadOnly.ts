import { RapiddocsOpenapiIr } from "@khulnasoft/openapi-ir";

export function shouldSkipReadOnly(method: RapiddocsOpenapiIr.HttpMethod): boolean | undefined {
    return method === "POST" || method === "PUT" || method === "PATCH";
}
