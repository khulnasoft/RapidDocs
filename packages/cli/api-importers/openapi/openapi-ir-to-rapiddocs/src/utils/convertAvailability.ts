import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Availability } from "@khulnasoft/openapi-ir";

export function convertAvailability(
    availability: Availability | undefined
): RawSchemas.AvailabilityUnionSchema | undefined {
    switch (availability) {
        case Availability.Deprecated:
            return "deprecated";
        case Availability.Beta:
            return "pre-release";
        case Availability.GenerallyAvailable:
            return "generally-available";
        default:
            return undefined;
    }
}
