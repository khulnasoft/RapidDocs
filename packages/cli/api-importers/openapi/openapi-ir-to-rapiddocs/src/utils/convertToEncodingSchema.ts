import { assertNever } from "@khulnasoft/core-utils";
import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Encoding } from "@khulnasoft/openapi-ir";

export function convertToEncodingSchema(encoding: Encoding): RawSchemas.EncodingSchema {
    switch (encoding.type) {
        case "protobuf":
            return {
                proto: {
                    type: encoding.typeName
                }
            };
        default:
            assertNever(encoding.type);
    }
}
