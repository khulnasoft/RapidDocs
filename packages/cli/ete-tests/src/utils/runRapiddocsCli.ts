import { Options } from "execa";
import path from "path";

import { loggingExeca } from "@khulnasoft/logging-execa";

export async function runRapiddocsCli(args: string[], options?: Options): Promise<loggingExeca.ReturnValue> {
    return loggingExeca(undefined, "node", [path.join(__dirname, "../../../cli/dist/dev/cli.cjs"), ...args], {
        ...options,
        env: {
            ...options?.env,
            RAPIDDOCS_TOKEN: process.env.RAPIDDOCS_ORG_TOKEN_DEV
        },
        doNotPipeOutput: options?.reject === false
    });
}
