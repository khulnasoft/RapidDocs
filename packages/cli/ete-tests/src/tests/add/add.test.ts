import { getDirectoryContents, getDirectoryContentsForSnapshot } from "@khulnasoft/fs-utils";

import { runRapiddocsCli } from "../../utils/runRapiddocsCli";
import { init } from "../init/init";

describe("rapiddocs add", () => {
    it("rapiddocs add <generator>", async () => {
        const pathOfDirectory = await init();

        const add = async (generator: string) => {
            await runRapiddocsCli(["add", generator], {
                cwd: pathOfDirectory
            });
        };

        await add("rapiddocsapi/rapiddocs-java-sdk");
        await add("rapiddocs-postman");

        expect(await getDirectoryContentsForSnapshot(pathOfDirectory)).not.toBeNull();
    }, 60_000);

    it("rapiddocs add <generator> --group sdk", async () => {
        const pathOfDirectory = await init();

        const add = async (generator: string, groupName: string) => {
            await runRapiddocsCli(["add", generator, "--group", groupName], {
                cwd: pathOfDirectory
            });
        };

        await add("rapiddocs-typescript", "typescript");

        expect(await getDirectoryContentsForSnapshot(pathOfDirectory)).toMatchSnapshot();
    }, 60_000);
});
