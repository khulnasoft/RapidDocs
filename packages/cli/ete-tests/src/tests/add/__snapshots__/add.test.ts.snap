// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rapiddocs add > rapiddocs add <generator> --group sdk 1`] = `
[
  {
    "contents": [
      {
        "contents": [
          {
            "contents": "name: api
error-discrimination:
  strategy: status-code
",
            "name": "api.yml",
            "type": "file",
          },
          {
            "contents": "# yaml-language-server: $schema=https://raw.githubusercontent.com/khulnasoft/rapiddocs/main/rapiddocs.schema.json

service:
  auth: false
  base-path: /movies
  endpoints:
    createMovie:
      docs: Add a movie to the database
      method: POST
      path: /create-movie
      request: CreateMovieRequest
      response: MovieId

    getMovie:
      docs: Retrieve a movie from the database based on the ID
      method: GET
      path: /{id}
      path-parameters:
        id: MovieId
      response: Movie
      errors:
        - MovieDoesNotExistError
      examples:
        # Success response
        - path-parameters:
            id: tt0111161
          response:
            body:
              id: tt0111161
              title: The Shawshank Redemption
              rating: 9.3
        # Error response
        - path-parameters:
            id: tt1234
          response:
            error: MovieDoesNotExistError
            body: tt1234

types:
  MovieId:
    type: string
    docs: The unique identifier for a Movie in the database

  Movie:
    properties:
      id: MovieId
      title: string
      rating:
        type: double
        docs: The rating scale out of ten stars

  CreateMovieRequest:
    properties:
      title: string
      rating: double

errors:
  MovieDoesNotExistError:
    status-code: 404
    type: MovieId
",
            "name": "imdb.yml",
            "type": "file",
          },
        ],
        "name": "definition",
        "type": "directory",
      },
      {
        "contents": "{
    "organization": "rapiddocs",
    "version": "0.0.0"
}",
        "name": "rapiddocs.config.json",
        "type": "file",
      },
      {
        "contents": "# yaml-language-server: $schema=https://schema.buildwithrapiddocs.dev/generators-yml.json
default-group: local
groups:
  local:
    generators:
      - name: rapiddocsapi/rapiddocs-typescript-node-sdk
        version: 0.39.3
        output:
          location: local-file-system
          path: ../sdks/typescript
  typescript:
    generators:
      - name: rapiddocsapi/rapiddocs-typescript
        version: 0.39.3
",
        "name": "generators.yml",
        "type": "file",
      },
    ],
    "name": "rapiddocs",
    "type": "directory",
  },
]
`;
